### ISSUE TITLE
[processor/lsminterval] Behaviour of timestamp for aggregated metrics

### ISSUE THREAD

**Issue by lahsivjar** [elastic/opentelemetry-collector-components#158]
LSM interval creates aggregated metrics without utilizing the timestamp of the incoming events. This behaviour is based on the upstream [intervalprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/intervalprocessor) on which the lsmintervalprocessor is based on.

OTOH, apm-aggregation uses truncated event timestamp (based on aggregation interval) for bucketing data. Due to this, old metrics are grouped in a separate bucket. This would mean that, aggregations for old spans would have a new timestamp in OTel.

**Comment by felixbarny** [comment]
Maybe this is something that should be configurable (event timestamp vs arrival timestamp).

If we truncate the event timestamp, we'll need to add another dimension for late arrivals. Otherwise, we could create another data point for the same timestamp for the same time series, leading to duplicate rejections.

This was one of the reasons that made it hard for use to adopt TSDB for APM.

**Comment by lahsivjar** [comment]
> If we truncate the event timestamp, we'll need to add another dimension for late arrivals. Otherwise, we could create another data point for the same timestamp for the same time series, leading to duplicate rejections.

Good point! It seems duplicate rejection would always be the case with APM logic for late arrivals which didn't make their realtime aggregation bucket. The current lsmintervalprocessor doesn't consider timestamp as an aggregation dimension but if we want to do it, we would have to truncate the event timestamp making data rejection a lot more probable. We could add another dimension, like a current truncated aggregation processing window which should make it unique but this would mean a new timeseries for every aggregation period... not sure if this is a good idea.

Alternatively, if we decide to leave things as they are i.e. aggregating with arrival timestamp then the resulting data might be a bit skewed. We could decide to only aggregate late arrivals up to a limit and minimize the deviation of the skew but not sure if that is any better.

**Comment by felixbarny** [comment]
What I'm thinking of is to add a numeric dimension (for example named `offset`) that indicates how "late" the data is. For example, if we aggregate spans for the current time, the offset would be `0`. For  the 1m bucket, if we aggregate a span from the past minute, the offset is `1`, and so on. So we "only" create a new time series for actual late arriving data.

While that would work well with delta temporality, I don't think it would really work with cumulative temporality.

**Comment by lahsivjar** [comment]
Hmm, that's a neat trick which gives me another idea. We could use the concept of `offset` but instead of encoding it as a separate attribute, we could encode it in the timestamp. If we were to emit event timestamp for aggregated metrics, we would truncate the UNIX timestamp as per a given interval. This would give us exact points in timestamp for a `0` offset aggregated metric. For any late arrivals, we can calculate its offset and add that many milliseconds to the timestamp. Even if an aggregation interval is 1 minute in length, we would be able to accommodate up to an hour's worth of late arrivals.

**Comment by felixbarny** [comment]
That won't allow us to accept timestamps that are arbitrarily late. But TSDB already has limitations on late-arriving data, so that's probably fine. Also, if we use nanosecond precision timestamps, it will be even less of an issue.

Maybe we should have a configurable set of strategies as the ideal strategy probably depends on the capabilities of the backend (nanosecond support, support for late arrivals, etc.)

**Comment by felixbarny** [comment]
Does https://github.com/elastic/opentelemetry-collector-components/pull/389 complicate things once we want to add support for using the signal's timestamp rather than the arrival timestamp?

**Comment by lahsivjar** [comment]
In my vision, we would have a max lookback time (set based on how old data we want to ingest). Any data older than this would not be ingested and we can use the max lookback time to increase the time range that we will need to iterate over. Does this make sense?

**Comment by lahsivjar** [comment]
This issue was discussed in the technical sync today. The behaviour of timestamps with the current approach (arrival timestamps) is not the best as it would lead to different results in the UI when using raw traces vs aggregated metrics. However, the priority of this is not clear and we also need more discussion on how to address the timestamps properly.

**Comment by lahsivjar** [comment]
The issue still persists but there have been some changes to where the issue will need to be fixed. Since the APM aggregations were broken into `signaltometricsconnector` and `lsmintervalprocessor` the issue now needs to be fixed in both the components.

### Signal to metrics connector

The `signaltometricsconnector` always uses `arrival` timestamp to produce metrics. https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/38602 upstream issue tracks this.

### LSM interval processor

LSM interval processor aggregates events in aggregation windows dervied from arrival time, however, within the window the data is aggregated based on event timestamp:

1. For delta temporality the latest observed metric is used to set the timestamp of the final metric.
2. For cumulative temporality the highest observed timestamp is used.

For overflows, the component does not set any timestamps, ideally we should be setting processing time as the timestamp for the overflow metrics.

### Summary of what needs to be done

1. `signaltometricsconnector` needs to support event timestamp. Tracked by https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/38602.
2. `lsmintervalprocessor` needs to support aggregating late arrivals in a way that we do not produce duplicate data for TSDB (see above comments for discussion around this).
3. [OPTIONAL] `lsmintervalprocessor` should also set timestamp for overflow metrics based on the processing time.

I am keeping this issue open as a `meta` issue on `elaticapmconnector` which wraps both LSM interval processor and signal to metrics connector.

**Comment by lahsivjar** [comment]
The issue still persists but there have been some changes to where the issue will need to be fixed. Since the APM aggregations were broken into `signaltometricsconnector` and `lsmintervalprocessor` the issue now needs to be fixed in both the components.

I am keeping this issue open as a `meta` issue on `elaticapmconnector` which wraps both LSM interval processor and signal to metrics connector. I have also updated the issue description to be based on the latest state of the components.
